### PR TITLE
Field index values as iterator instead of slice

### PR DESCRIPTION
**Pull request by IvanPleshkov**
Tracking issue: #4502.

Integration of this PR https://github.com/qdrant/qdrant/pull/4641.

This PR changes signature of `[ImmutablePointToValues::get_values](https://github.com/qdrant/qdrant/blob/field-index-values-as-iterator-instead-of-slice/lib/segment/src/index/field_index/immutable_point_to_values.rs#L36)`

We cannot use slice anymore because we cannot store slice of strings in mmap file. Instead, `get_values` returns values iterator.

### All Submissions:

* [x] Contributions should target the `dev` branch. Did you create your branch from `dev`?
* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### New Feature Submissions:

1. [x] Does your submission pass tests?
2. [x] Have you formatted your code locally using `cargo +nightly fmt --all` command prior to submission?
3. [x] Have you checked your code using `cargo clippy --all --all-features` command?

### Changes to Core Features:

* [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your core changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

File: lib/segment/src/index/field_index/map_index/immutable_map_index.rs [hunk 1]
```diff
@@ -137,7 +133,16 @@ impl<N: Hash + Eq + Clone + Display + FromStr + Default> ImmutableMapIndex<N> {
...
+                removed_values_count += 1;
+            }
+
+            if removed_values_count > 0 {
+                self.indexed_points -= 1;
             }
+            self.values_count = self
+                .values_count
+                .checked_div(removed_values_count)
+                .unwrap_or_default();
```
Comment (timvisee) [inline lib/segment/src/index/field_index/map_index/immutable_map_index.rs:145]: I don't think this matches with `self.values_count -= removed_values.len();` from before. Is that intentional?

**Comment by xzfc** [comment]
I see you've added `get_values_count()` method. Have you considered implementing `ExactSizeIterator` instead?

**Review by timvisee — APPROVED** [review]
Looks good once we fix the current open issues.

**Comment by IvanPleshkov** [comment]
> I see you've added `get_values_count()` method. Have you considered using `ExactSizeIterator` instead?

PR with point-to-values map https://github.com/qdrant/qdrant/pull/4641/files cannot implement `ExactSizeIterator` because in mmap case we scan values in only forward direction

File: lib/segment/src/index/field_index/map_index/immutable_map_index.rs (hunk 1 above)
Comment (IvanPleshkov) [reply lib/segment/src/index/field_index/map_index/immutable_map_index.rs]: thanks! it's a bug

**Pull request merged**